cmd/encryption-v1.go

```diff
@@ -161,31 +181,44 @@ func EncryptRequest(content io.Reader, r *http.Request, metadata map[string]stri
 	delete(metadata, SSECustomerKey) // make sure we do not save the key by accident
 
 	// security notice:
-	// Reusing a tuple (nonce, client provided key) will produce the same encryption key
-	// twice and breaks the tamper-proof property. However objects are still confidential.
-	// Therefore the nonce must be unique but need not to be undistinguishable from true
-	// randomness.
-	nonce := make([]byte, 32) // generate random nonce to derive encryption key
+	//  - If the first 32 bytes of the random value are ever repeated under the same client-provided
+	//    key the encrypted object will not be tamper-proof. [ P (coll) ~= 1 / 2^(256 / 2)]
+	//  - If the last 32 bytes of the random value are ever repeated under the same client-provided
+	//    key an adversary may be able to extract the object encryption key. This depends on the
+	//    authenticated en/decryption scheme. The DARE format will generate an 8 byte nonce which must
+	//    be repeated in addition to reveal the object encryption key.
+	//    [ P[coll] ~= 1 / 2^((256 + 64) / 2) ]
+	nonce := make([]byte, 64) // generate random values for key derivation
 	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
 		return nil, err
 	}
-	iv := sha256.Sum256(nonce) // hash output to not reveal any stat. weaknesses of the PRNG
+	sha := sha256.New() // derive object encryption key
+	sha.Write(key)
+	sha.Write(nonce[:32])
+	objectEncryptionKey := sha.Sum(nil)
 
-	keyHash := sha256.Sum256(key) // derive MAC of the client-provided key
-	mac := hmac.New(sha256.New, keyHash[:])
-	mac.Write(iv[:])
-	keyMAC := mac.Sum(nil)
+	iv := sha256.Sum256(nonce[32:]) // derive key encryption key
+	sha = sha256.New()
+	sha.Write(key)
+	sha.Write(iv[:])
+	keyEncryptionKey := sha.Sum(nil)
 
-	mac = hmac.New(sha256.New, key) // derive encryption key
-	mac.Write(keyMAC)
-	reader, err := sio.EncryptReader(content, sio.Config{Key: mac.Sum(nil)})
+	sealedKey := bytes.NewBuffer(make([]byte, 0, 64)) // 16 byte header + 32 byte payload + 16 byte tag
```

Comment:
could declare (fixed length buffer of 64 bytes) on stack

Reply:
Tried that before but the growing functionality of `bytes.Buffer` allocates a new buffer 
(yes, if I declare:
```
var sealedKey [64]byte
n, err := sio.Encrypt(bytes.NewBuffer(sealedKey[:0]), ...
```
)

I can explore why / reason...
The problem is that if we use for example:
```
var sealedKey [64]byte // or any other size instead of 64
n, err := sio.Encrypt(bytes.NewBuffer(sealedKey[:0]), ...)
...

metadata[ServerSideEncryptionSealedKey] = base64.StdEncoding.EncodeToString(sealedKey[:]) // or sealedKey[:64]  
```
and the `bytes.Buffer` made heap copy because of growing than `sealedKey[:]` contains zeros and object cannot be decrypted. So we would depend on a non-obvious implementation detail here. That can cause horrible bugs IMO... 
```
Nevertheless I try to figure out why `bytes.Buffer` grows because from my understanding there should be no reason...

Reply:
Okay, reason is:
1. bytes.Buffer implements `io.ReadFrom`. `io.Copy` (which is used by `sio.Encrypt`) prefers
   `ReadFrom` function of an `io.Writer` over doing the copy with a buffer.
2. bytes.Buffer's `ReadFrom` implementation allocates buffer always if remaining cap is smaller than 512. See [ReadFrom](https://golang.org/src/bytes/buffer.go?s=6465:6524#L197)

Reply:
Bytes buffer implementation is meant to be grown and it will grow with your control 2x times.

Reply:
Yes, but you "cannot" predict whether it grows. For example the given slice should have enough capacity to hold all data -> no need to grow but the `ReadFrom` function grow always if `cap - len` <  512 even if there would be enough space.

If you write: 
```
buffer := bytes.NewBuffer(make([]byte, 0, 100))
reader := bytes.NewReader(make([]byte, 99))
io.Copy(buffer, reader)
```
the buffer will still do a grow even though it would not be necessary.

Reply:
Ok, fine.